src/main/kotlin/app/CodeLongevity.kt

```diff
@@ -0,0 +1,266 @@
+                for (edit in editList) {
+                    val delStart = edit.getBeginA()
+                    val delEnd = edit.getEndA()
+                    val delCount = edit.getLengthA()
+                    var insStart = edit.getBeginB()
+                    var insEnd = edit.getEndB()
+                    val insCount = edit.getLengthB()
+                    println("del ($delStart, $delEnd), ins ($insStart, $insEnd)");
+
+                    // Deletion case. Chase down the deleted lines through the
+                    // history.
+                    if (delCount > 0) {
+                        var tmpLines = ArrayList<RevCommitLine>(delCount)
+                        var idx = delStart;
+                        while (idx < delEnd) {
+                            tmpLines.add(RevCommitLine(commit, oldPath, idx));
+                            idx++
+                        }
+
+                        lines.addAll(delStart, tmpLines)
+                    }
+
+                    // Insertion case. Report it.
+                    if (insCount > 0) {
+                        val fileLoader = repo.open(newId)
+                        val fileText = RawText(fileLoader.getBytes())
+
+                        for (idx in insStart .. insEnd - 1) {
+                            val from = RevCommitLine(commit, newPath, idx)
+                            var to = lines.get(idx)
+                            val cl = CodeLine(from, to, fileText.getString(idx))
+                            codeLines.add(cl)
+                        }
+                        lines.subList(insStart, insEnd).clear()
+                    }
+                }
+
+                // File was renamed, tweak the files map.
+                if (diff.changeType == DiffEntry.ChangeType.RENAME) {
+                    files.set(oldPath, files.remove(newPath)!!)
+                }
+            }
+            commit = parentCommit
+        }
+        while (parentCommit != null && parentCommit != tail)
+
+        // If a tail revision was given then the map has to contain unclaimed
+        // code lines, i.e. the lines added before the tail revision. Push
+        // them all into the result lines list, so the caller can update their
+        // ages properly.
+        if (tail != null) {
+            for ((file, lines) in files) {
+                for (idx in 0 .. lines.size - 1) {
+                    val from = RevCommitLine(tail, file, idx)
+                    val cl = CodeLine(from, lines[idx], "no data (too lazy to compute)")
+                    codeLines.add(cl)
+                }
+            }
+        }
+    }
+
+    /**
+     * Returns lines array of a text file of the given revision.
+     */
+    private fun getFileLines(fileId: ObjectId, filePath: String,
+                             commit: RevCommit) : ArrayList<RevCommitLine>? {
+        val fileLoader = repo.open(fileId)
+        if (RawText.isBinary(fileLoader.openStream())) {
+            return null
```

Comment:
It feels more complicated with me, null is sort of convenient way to report about error. I'm good with empty array as a result, however I would say the code will be less efficient. I don't think though it will be ever a bottleneck. So if you really prefer your way, then I can do it, otherwise I would keep it as is.